src/types/node.rs

```diff
@@ -0,0 +1,238 @@
+use std::sync::Arc;
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use super::util::*;
+use super::*;
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+pub struct Node {
+    pub node_id: NodeId,
+    pub input_pads: Option<Vec<NodeId>>,
+    pub fallback_id: Option<NodeId>,
+
+    #[serde(flatten)]
+    pub params: NodeParams,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
+pub enum NodeParams {
+    WebRenderer(WebRenderer),
+    Shader(Shader),
+    Image(Image),
+    Text(Text),
+    Transition(Transition),
+    #[serde(rename = "builtin:transform_to_resolution")]
+    TransformToResolution(TransformToResolution),
+    #[serde(rename = "builtin:fixed_position_layout")]
+    FixedPositionLayout(FixedPositionLayout),
+    #[serde(rename = "builtin:tiled_layout")]
+    TiledLayout(TiledLayout),
+    #[serde(rename = "builtin:mirror_image")]
+    MirrorImage(MirrorImage),
+    #[serde(rename = "builtin:corners_rounding")]
+    CornersRounding(CornersRounding),
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct WebRenderer {
+    pub instance_id: RendererId,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct Image {
+    pub image_id: RendererId,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct Shader {
+    pub shader_id: RendererId,
+    pub shader_params: Option<ShaderParam>,
+    pub resolution: Resolution,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+#[serde(
+    tag = "type",
+    rename_all = "snake_case",
+    content = "value",
+    deny_unknown_fields
+)]
+pub enum ShaderParam {
+    F32(f32),
+    U32(u32),
+    I32(i32),
+    List(Vec<ShaderParam>),
+    Struct(Vec<ShaderParamStructField>),
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
+pub struct ShaderParamStructField {
+    pub field_name: String,
+    #[serde(flatten)]
+    pub value: ShaderParam,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema)]
```

Comment:
One additional approach I considered was to e.g. wrap all related types into mod. In this case it would be Text,TextStyle,TextWrapMode,TextWeight, TextDimensions

- It would group related types together (from_node.rs is already doing it, but inside wrap function)
- We could consider using serde defaults in that case because all those default_stomehting functions would be encapsulted inside the module
- If we use serde defaults that info would be included inside json schema